src/lib.rs

```diff
@@ -855,9 +855,10 @@ impl<T> Async<T> {
     /// inner.set_nonblocking(false)?;
     /// # std::io::Result::Ok(()) });
     /// ```
-    pub fn into_inner(mut self) -> io::Result<T> {
-        let io = self.io.take().unwrap();
-        Reactor::get().remove_io(&self.source)?;
+    pub fn into_inner(self) -> io::Result<T> {
+        let Self { mut source, io } = self;
+        source.remove()?;
+        std::mem::forget(source);
```

Comment:
This leaks the `Arc` and everything inside. I think you have to either:

1). Make `SourceContainer` a wrapper around `Option<Arc<Source>>` and have `remove()` take it out.
2). Make `Async` just contain `Option<SourceContainer>`.

Reply:
It appears that satisfying these two prerequisites proves to be an impasse: 

1. Bypassing the usage of `Option`
2. Preserving the semantics of the method `Async::into_inner`

Therefore, the only logical step forward is to close this pull request. I appreciate all your reviews.